### DISCUSSION TITLE
VPNs and share issues (port issues)

### DISCUSSION THREAD

**Discussion by atomGit** [nicotine-plus/nicotine-plus#2622] [General]
many people use VPNs these days and in such cases it's often not possible or advisable to forward ports and this creates problems for the network regarding sharing/browsing files

can anything be done about this?

i'm guessing that any solution would have to come from upstream (soulseek), but i don't know that for sure

**Comment by slook** [comment]
> can anything be done about this?

In theroey you might run something like pfSense on a remotely hosted virtual machine, and then try to configure some sort of complex reverse proxy forwarder relay to route peer connections to your local client. No easy solutions currently exist for doing that, however.

At that point, you might as well run a client in a remote Docker container, and then connect to the slsk network from there instead...

...There are [possible solutions](https://github.com/nicotine-plus/nicotine-plus/discussions/2312) for this that do exist for [Nicotine+](https://github.com/cproensa/nicotine-docker#nicotine-novnc-docker-container), but the [slskd](https://github.com/slskd/slskd) client is much better suited for that purpose, because it has a web-based interface.

In any case, you would require admin access to a hosted VPS, rather than a VPN, to do anything like as described above.

**Comment by mathiascode** [comment]
The options nowadays, pretty much:
- Use a VPN that supports port forwarding (which are becoming rare, unfortunately)
- Use a server (e.g. VPS) with a public IPv4 address
- Look into TCP hole punching, and see if this can be done with the current Soulseek protocol (but it's not going to be 100% reliable in any case, and I'm not sure to what extent TCP hole punching works in practise)

**Comment by atomGit** [comment]
i ended up forwarding a port on my VPN, though only reluctantly

just a note to whoever may run across this - virtually all VPNs [are shit](https://odysee.com/@NaomiBrockwell:4/vpn-scandals:5) - there are only a small handful i know of that are worth considering; mullvad, ovpn and azirevpn (another may be ivpn) - all of which claim to physically own and secure their servers (only a small portion in mullvad's case) and all of which allow port forwarding to my knowledge

**Reply by mathiascode** [reply]
Mullvad and IVPN removed support for port forwarding.

https://mullvad.net/en/blog/2023/5/29/removing-the-support-for-forwarded-ports/
https://www.ivpn.net/blog/gradual-removal-of-port-forwarding/

**Comment by atomGit** [comment]
ah, that's right - i forgot about that and i read the mullvad article recently too

**Comment by DeathStalker77** [comment]
Is it just the 3.3.0 B5 Nightly build that does not allow the updating of the Listening Port?

No matter what I do, it will NOT change, and it is ALWAYS registered as CLOSED (even with no firewall).  This is AFTER the re-start that has caused SO many problems :-(

I would _expect_ that, if I select a Listening Port and hit APPLY, then what is shown for the **Public UP Address** should be immediately updated to reflect the Listening Port that was chosen.

--- DS

**Reply by mathiascode** [reply]
Changing your port requires either a restart or reconnecting to the server (Menu -> Disconnect, then Menu -> Connect). We should probably make that clearer.

I could also make Nicotine+ reconnect automatically after changing the port, but I'm concerned if this would be unexpected and disrupt any transfer queues going on.

**Reply by mathiascode** [reply]
By the way, can you try this latest development build and see if anything is different? https://nightly.link/mathiascode/nicotine-plus/actions/runs/6450996981

**Comment by DeathStalker77** [comment]
Ah!  Ok, yes, that disconnect-reconnect was not clear at all :-)  I agree that _just_ an automatic process could be disruptive, but why not just add the option for either one?  Have the user choose whether they want it to automatically change or to have to manually disconnect-reconnect (one-button option for that would be my suggestion).

I will try the new build (x64 installer) as soon as possible :-)

--- DS

**Reply by slook** [reply]
Another possibility for improved design might be that certain entry fields, such as port number and username, ought to be disabled (greyed-out) if connected, with a tooltip explaining this setting cannot be altered while online.

**Reply by mathiascode** [reply]
> Another possibility for improved design might be that certain entry fields, such as port number and username, ought to be disabled (greyed-out) if connected, with a tooltip explaining this setting cannot be altered while online.

There's the risk that people don't know how to disconnect manually, and think something is broken.

It will require some experimentation, but I don't think we have time for the 3.3.0 release. I'm thinking about just adding the `(requires restart)` text back for now, even though it technically isn't true anymore.

**Comment by DeathStalker77** [comment]
FWIW, PIA, AirVPN, VyperVPN, & Proton VPN all support Port Forwarding, though currently only a couple of them support multiple Ports - they do that by opening ALL ports and you are provided with a static IP (which, personally, I'm not thrilled with).  I am still waiting for confirmation that Proton VPN does indeed support multiple ports (I have heard it from users, but it is not indicated on their site).  PVPN would be the one I would trust the most, even though it is more expensive than the others, and does not have as many servers in as many locations as other VPNs.

**Comment by xristos10** [comment]
I use AirVPN it offers more than 1 port for forwarding that they don't randomize each time it activates, you can have soulseek/nicotine+ and a torrent client working together at the same time (each program needs its own dedicated port*), and finally you also have the option to manually set split tunneling (I use WireSockUI for that instead of the officlal client) so you can set which programs will pass through vpn and which are going to use your normal connection.
AirVPN devs are supposed to work on implementing split tunneling on their client but this may take years

*most VPNs that offer port forwarding give only 1 port, so you can only have one program using it.

1# This is not an ad but an answer with a solution, I suggest that if you decide to give it a shot then buy a subscription for a few days first to see if it meets your needs, before moving to annual etc.
2# If you happen to read this comment months from now don t take it for granted that it's up to date, check it on your own  : )

**Reply by DeathStalker77** [reply]
I have been considering AirVPN.  PIA is _supposed_ to be considering extra ports for an additional cost.

**Reply by xristos10** [reply]
I think AirVPN currently offers 5 forwarded ports per new user.
